File: java/po/world_simulator/entities/Animal.java
package po.world_simulator.entities;

import po.world_simulator.Entity;
import po.world_simulator.Map;
import po.world_simulator.Vector2;
import po.world_simulator.World;

public abstract class Animal extends Entity {
    private boolean canKill;

    public Animal(int power, int initiative, Vector2 position, String name, String datatype, int age, boolean canKill) {
        super(power, initiative, position, name, datatype, age);
        this.canKill = canKill;
    }

    public Animal(int power, int initiative, Vector2 position, String name, String datatype) {
        this(power, initiative, position, name, datatype, 0, false);
    }

    @Override
    public void action() {
        // [AI] Wywoływane akcje - do implementacji wg konkretnej logiki c++ (losowy kierunek itp)
        Vector2 moveDirection = World.getMap().getPossibleMoveDirection(this.position, false);
        if (!moveDirection.equals(Vector2.ZERO)) {
            move(moveDirection);
        }
        this.age += 1;
    }

    @Override
    public void collision(Entity otherEntity) {
        if (this.getClass().equals(otherEntity.getClass())) {
            this.reproduce();
            return;
        }

        if (otherEntity instanceof Animal) {
            Animal enemy = (Animal) otherEntity;
            if (this.getPower() <= enemy.getPower() && !enemy.isPredator()) {
                this.runAway();
                return;
            }
        }
        super.collision(otherEntity);
    }

    public boolean isPredator() {
        return canKill;
    }

    protected void move(Vector2 moveDirection) {
        Map map = World.getMap();
        Vector2 targetPos = this.position.add(moveDirection);
        Entity collisionTarget = map.getEntityAt(targetPos);

        if (collisionTarget != null) {
            collisionTarget.collision(this);

            if (!this.isAlive()) {
                return;
            }
        }

        if (map.getEntityAt(targetPos) == null) {
            map.move(this.position, targetPos);
            this.position = targetPos;
        }
    }

    protected boolean runAway() {
        Map map = World.getMap();
        Vector2 direction = map.getPossibleMoveDirection(this.position, true);

        if (!moveDir.equals(Vector2.ZERO)) {
            move(direction);
            return true;
        }
        return false;
    }
}

File: python/po/world_simulator/entities/Animal.py
from .Entity import Entity
from po.world_simulator.Vector2 import Vector2
import po.world_simulator.World as world_module

class Animal(Entity):
    def __init__(self, power, initiative, position, name, age=0, canKill=False):
        super().__init__(power, initiative, position, name, age)
        self.canKill = canKill

    def action(self):
        moveDirection = world_module.World.getMap().getPossibleMoveDirection(self.position, False)
        if moveDirection != Vector2.ZERO:
            self.move(moveDirection)
        self.age += 1

    def collision(self, otherEntity):
        if type(self) == type(otherEntity):
            self.reproduce()
            return

        if isinstance(otherEntity, Animal):
            enemy = otherEntity
            if self.getPower() <= enemy.getPower() and not enemy.isPredator():
                self.runAway()
                return

        super().collision(otherEntity)

    def isPredator(self):
        return self.canKill

    def move(self, moveDirection):
        game_map = world_module.World.getMap()
        targetPos = self.position.add(moveDirection)
        collisionTarget = game_map.getEntityAt(targetPos)

        if collisionTarget is not None:
            collisionTarget.collision(self)

            if not self.isAlive():
                return

        if game_map.getEntityAt(targetPos) is None:
            game_map.move(self.position, targetPos)
            self.position = targetPos

    def runAway(self):
        game_map = world_module.World.getMap()
        direction = game_map.getPossibleMoveDirection(self.position, True)

        if direction != Vector2.ZERO:
            self.move(direction)
            return True
        return False